### ISSUE TITLE
Why coredump twice every time crashed

### ISSUE THREAD

**Issue by xujintao** [vmangos/core#220]
**Current behavior**: 
coredump twice
**Expected behavior**: 
coredump once, after all the file is big(about 1G)
**Steps to reproduce**:
```
case SIGSEGV:
    signal(SIGSEGV, 0);
    if (!m_handleSigvSignals)
        return;
    m_handleSigvSignals = false; // Desarm anticrash
    sWorld.SetAnticrashRearmTimer(sWorld.getConfig(CONFIG_UINT32_ANTICRASH_REARM_TIMER));
    uint32 anticrashOptions = sWorld.getConfig(CONFIG_UINT32_ANTICRASH_OPTIONS);
    // Log crash stack
    sLog.outInfo("Received SIGSEGV");
    ACE_Stack_Trace st;
    sLog.outInfo("%s", st.c_str());
    if (anticrashOptions & ANTICRASH_GENERATE_COREDUMP)
        createdump();
    if (anticrashOptions & ANTICRASH_OPTION_ANNOUNCE_PLAYERS)
    {
        if (anticrashOptions & ANTICRASH_OPTION_SAVEALL)
            sWorld.SendWorldText(LANG_SYSTEMMESSAGE, "Server has crashed. Now saving online players ...");
        else
            sWorld.SendWorldText(LANG_SYSTEMMESSAGE, "Crash server occurred :(");
        ACE_Based::Thread::Sleep(500);
    }
    if (anticrashOptions & ANTICRASH_OPTION_SAVEALL)
    {
        CharacterDatabase.ThreadStart();
        sObjectAccessor.SaveAllPlayers();
        ACE_Based::Thread::Sleep(25000); // Wait enough time to execute the SQL queries.
    }
    *((int*)NULL) = 42; // Crash for real now.
    return;
```
First time, coredump in child process with ANTICRASH_GENERATE_COREDUMP option on.
Second time, *((int*)NULL) = 42; // Crash for real now.

**Client version**: 1.12.1

**Commit hash**: 5166b62

**Operating system**: linux

**Comment by coolzoom** [comment]
the first time is SIGSEGV handler so it will catch the signal and save player and announce players before core. that is safer at production enviroment when there're many players online so the data is saved. the second one is a wrong equation intentionally to "crash for real", you could remove the function you dont need for your own purpose.

**Comment by xujintao** [comment]
1, Child handle dump, parent save field session to db and announce players.
2, After remove statement "*((int*)NULL) = 42; // Crash for real now."
the parent return directly, can the Master::Run() exit gracefully?

**Comment by coolzoom** [comment]
no. you ether use current method, or write your own signal handler.

**Comment by xujintao** [comment]
1, use current method.
Current default value of Anticrash.Options in mangosd.conf is
```
Anticrash.Options = 28
```
So, change it to 12 or 0 can avoid child dump.

2, other method(maybe a better idea)
Still let child to dump core, but parent do gracefully exit from Master::Run as SIGINT or SIGTERM instead of dump directly.

**Comment by ulfgebhardt** [comment]
@xujintao could you document this in the distributed config file so others can benefit from your findings. Furthermore would you open an PR for your suggested change?

**Comment by punker85** [comment]
If this is still an issue, you may want to remove the logic at the end of the handler - `*((int*)NULL) = 42; // Crash for real now.`. This line should be causing a segmentation fault all over again, which should cause the handler to be re-entered once again and core dump again.

**Comment by ulfgebhardt** [comment]
@ratkosrb Can you have a look at the suggestion in the last comment pls? <3

**Comment by punker85** [comment]
> @ratkosrb Can you have a look at the suggestion in the last comment pls? <3

The problem I suggested is more than likely not the issue but it probably does not help. In this section of code, the likely culprit is the 25s thread sleep. In my experience with signal handling, ignoring the signal for extended periods of time will cause the OS to retry execution of the SEGV-causing code (or possibly send a new termination signal), which would cause the process to handle the signal in the default manner as prescribed by signal(SIGSEGV, 0).

If you read the answer at this link (https://stackoverflow.com/questions/8456085/why-cant-i-ignore-sigsegv-signal), it claims the following will happen after catching and handling the signal: "The kernel restarts the offending operation"
